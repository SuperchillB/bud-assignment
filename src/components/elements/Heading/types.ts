import {
  BorderProps,
  ColorProps,
  FlexboxProps,
  LayoutProps,
  PositionProps,
  SpaceProps,
  TypographyProps,
} from 'styled-system';

export type HeadingFunctionsProps = SpaceProps &
  ColorProps &
  LayoutProps &
  FlexboxProps &
  BorderProps &
  PositionProps &
  TypographyProps;

export type HeadingProps = {
  variant: 'h2' | 'h3'; // Could add more headings (h1, h4, ...)
} & HeadingFunctionsProps;
