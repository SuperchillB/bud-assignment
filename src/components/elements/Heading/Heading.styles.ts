import styled from 'styled-components';
import {
  border,
  color,
  compose,
  flexbox,
  layout,
  position,
  space,
  typography,
  variant,
} from 'styled-system';
import { HeadingProps } from './types';

const headingStyles = {
  mt: '0rem',
  mb: '0rem',
  color: 'midnight',
  fontWeight: 'bold',
};

export const Heading = styled.h1.attrs<HeadingProps>(({ variant: size }) => ({
  as: size,
}))<HeadingProps>(
  variant({
    variants: {
      h1: {
        ...headingStyles,
        fontSize: 'xxlarge',
      },
      h2: {
        ...headingStyles,
        fontSize: 'xlarge',
      },
      h3: {
        ...headingStyles,
        fontSize: 'large',
      },
      h4: {
        ...headingStyles,
        fontSize: 'medium',
      },
      // Could add more headings (h5, ...)
    },
  }),
  compose(space, color, layout, flexbox, border, position, typography),
);

Heading.defaultProps = {
  variant: 'h2',
};
