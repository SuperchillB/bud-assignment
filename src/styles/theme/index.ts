import radii from './foundations/border';
import breakpoints from './foundations/breakpoints';
import { colors } from './foundations/colors';
import shadows from './foundations/shadows';
import space from './foundations/space';
import { fonts, fontSizes, fontWeights } from './foundations/typography';
import { ThemeType } from './types';

const theme: ThemeType = {
  breakpoints,
  colors,
  fonts,
  fontSizes,
  fontWeights,
  radii,
  shadows,
  space,
};

export default theme;
