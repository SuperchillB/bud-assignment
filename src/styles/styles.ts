import { createGlobalStyle } from 'styled-components';
import { normalize } from 'styled-normalize';
import { ThemeType } from './theme/types';

export const GlobalStyles = createGlobalStyle<{ theme: ThemeType }>`
  ${normalize};
  *, :before, :after {
    box-sizing: border-box;
    margin: 0;
    padding: 0;
  }
  html {
    font-size: 62.5%;
    scroll-behavior: smooth;
  }
  body {
    /* font-family: ${(props) => props.theme.fonts.primary}; */
    font-family: ${({ theme }) => theme.fonts.primary};
    font-size: 1.6rem;
    background: ${({ theme }) => theme.colors.info};
    color: ${(props) => props.theme.colors.primary}; // ! CHANGE LATER
    cursor: default;
  }
  main {
    height: 100%;
  }
  h1,h2,h3,h4,h5,h6,button {
    font-family: ${(props) => props.theme.fonts.title}; // ! CHANGE LATER
    line-height: 2rem;
  }
  p {
    line-height: 2rem;
  }
  h1 {
    font-weight: normal;
    font-size: 3.2rem;
  }
  a {
    color: inherit;
    text-decoration: none;
  }
  button {
    border: 0;
    padding: 0;
    background: none;
    cursor: pointer;
  }
  li{
    list-style: none;
  }
`;
