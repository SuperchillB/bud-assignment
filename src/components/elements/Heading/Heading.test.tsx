import { render, screen as screenRTL } from '@testing-library/react';
import { Heading } from './Heading.styles';

describe('Heading component', () => {
  test('renders its text and id', () => {
    const id = 'testId';
    const children = 'Test heading';
    const { container } = render(
      <Heading id={id} variant="h2">
        {children}
      </Heading>,
    );

    expect(container.querySelector('h2')?.id).toEqual(id);
    expect(screenRTL.getByText(children)).toBeInTheDocument();
  });

  test('renders with given props and "h3" variant', () => {
    const { container } = render(<Heading variant="h3" />);

    const h3 = container.querySelector('h3');
    expect(h3).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders with given props and "h4" variant', () => {
    const { container } = render(<Heading variant="h4" />);

    const h4 = container.querySelector('h4');
    expect(h4).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders with given props', () => {
    const { container } = render(
      <Heading
        alignItems={['center', 'stretch']}
        backgroundColor="budLightBlue"
        color="primary"
        flexDirection="column"
        justifyContent="center"
        mb={['medium', 'large', 'xlarge']}
        px="small"
        width={1}
        variant="h3"
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
