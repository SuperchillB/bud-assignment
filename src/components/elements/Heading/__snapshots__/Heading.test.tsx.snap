// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading component renders with given props 1`] = `
<h3
  class="sc-bczRLJ jA-dmtd"
  color="primary"
  width="1"
/>
`;

exports[`Heading component renders with given props and "h3" variant 1`] = `
<h3
  class="sc-bczRLJ bAcZEu"
/>
`;

exports[`Heading component renders with given props and "h4" variant 1`] = `
<h4
  class="sc-bczRLJ gzwZGm"
/>
`;
